[Bug]: datacoord crashes when segment deleted which some legacy index

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: d0b974d
- Deployment mode(standalone or cluster): both
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```


### Current Behavior

`DataCoord` may panic if Milvus restart with segment info deleted while index info left.

### Expected Behavior

Datacoord shall not panic in this case

### Steps To Reproduce

_No response_

### Milvus Log

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x24fc190]

goroutine 152 [running]:
panic({0x33555c0, 0x4e3f010})
	/usr/local/go/src/runtime/panic.go:941 +0x397 fp=0xc000e57580 sp=0xc000e574c0 pc=0x11e3c17
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:220
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:818 +0x31d fp=0xc000e575d0 sp=0xc000e57580 pc=0x11fabdd
github.com/milvus-io/milvus/internal/datacoord.(*SegmentsInfo).SetSegmentIndex(...)
	/home/silverxia/workspace/milvus/internal/datacoord/segment_info.go:102
github.com/milvus-io/milvus/internal/datacoord.(*meta).updateSegmentIndex(...)
	/home/silverxia/workspace/milvus/internal/datacoord/index_meta.go:45
github.com/milvus-io/milvus/internal/datacoord.(*meta).reloadFromKV(0xc0001b62a0)
	/home/silverxia/workspace/milvus/internal/datacoord/meta.go:147 +0x970 fp=0xc000e57840 sp=0xc000e575d0 pc=0x24fc190
github.com/milvus-io/milvus/internal/datacoord.newMeta({0x3ab91b0?, 0xc000efbec0}, {0x3addf10?, 0xc0001918f0}, {0x3ad7c10?, 0xc000191860})
	/home/silverxia/workspace/milvus/internal/datacoord/meta.go:92 +0x1dd fp=0xc000e578b0 sp=0xc000e57840 pc=0x24fb79d
github.com/milvus-io/milvus/internal/datacoord.(*Server).initMeta.func1()
	/home/silverxia/workspace/milvus/internal/datacoord/server.go:474 +0x127 fp=0xc000e57930 sp=0xc000e578b0 pc=0x251c367
github.com/milvus-io/milvus/internal/util/retry.Do({0x3ab91b0, 0xc000efbec0}, 0xc000e57b28, {0xc000e53b20, 0x1, 0x3ab91b0?})
	/home/silverxia/workspace/milvus/internal/util/retry/retry.go:37 +0x103 fp=0xc000e57ac8 sp=0xc000e57930 pc=0x1dbb783
github.com/milvus-io/milvus/internal/datacoord.(*Server).initMeta(0xc00025e380, {0x3ad7c10, 0xc000191860})
	/home/silverxia/workspace/milvus/internal/datacoord/server.go:480 +0x176 fp=0xc000e57b60 sp=0xc000e57ac8 pc=0x251c1d6
github.com/milvus-io/milvus/internal/datacoord.(*Server).Init(0xc00025e380)
	/home/silverxia/workspace/milvus/internal/datacoord/server.go:281 +0xbf fp=0xc000e57c70 sp=0xc000e57b60 pc=0x2519d3f
github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).init(0xc000f2a060)
	/home/silverxia/workspace/milvus/internal/distributed/datacoord/service.go:107 +0x502 fp=0xc000e57ee8 sp=0xc000e57c70 pc=0x2540a82
github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).Run(0xb?)
	/home/silverxia/workspace/milvus/internal/distributed/datacoord/service.go:215 +0x25 fp=0xc000e57f28 sp=0xc000e57ee8 pc=0x2542145
github.com/milvus-io/milvus/cmd/components.(*DataCoord).Run(0xc0006d7c00?)
	/home/silverxia/workspace/milvus/cmd/components/data_coord.go:49 +0x1d fp=0xc000e57f60 sp=0xc000e57f28 pc=0x2c7e0fd
github.com/milvus-io/milvus/cmd/roles.runComponent[...].func1()
	/home/silverxia/workspace/milvus/cmd/roles/roles.go:89 +0x150 fp=0xc000e57fe0 sp=0xc000e57f60 pc=0x2c812f0
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc000e57fe8 sp=0xc000e57fe0 pc=0x1219001
created by github.com/milvus-io/milvus/cmd/roles.runComponent[...]
	/home/silverxia/workspace/milvus/cmd/roles/roles.go:76 +0x12f
```

### Anything else?

_No response_

--- a/internal/datacoord/index_meta_test.go
+++ b/internal/datacoord/index_meta_test.go
@@ -1234,3 +1234,21 @@ func TestMeta_GetHasUnindexTaskSegments(t *testing.T) {
 		assert.Equal(t, segID, segments[0].ID)
 	})
 }
+
+// see also: https://github.com/milvus-io/milvus/issues/21660
+func TestUpdateSegmentIndexNotExists(t *testing.T) {
+	m := &meta{
+		segments: &SegmentsInfo{
+			segments: map[UniqueID]*SegmentInfo{},
+		},
+		indexes:              map[UniqueID]map[UniqueID]*model.Index{},
+		buildID2SegmentIndex: make(map[UniqueID]*model.SegmentIndex),
+	}
+
+	assert.NotPanics(t, func() {
+		m.updateSegmentIndex(&model.SegmentIndex{
+			SegmentID: 1,
+			IndexID:   2,
+		})
+	})
+}

--- a/internal/datacoord/segment_info.go
+++ b/internal/datacoord/segment_info.go
@@ -19,7 +19,9 @@ package datacoord
 import (
 	"time"
 
+	"github.com/milvus-io/milvus/internal/log"
 	"github.com/milvus-io/milvus/internal/metastore/model"
+	"go.uber.org/zap"
 
 	"github.com/golang/protobuf/proto"
 	"github.com/milvus-io/milvus-proto/go-api/commonpb"
@@ -99,10 +101,18 @@ func (s *SegmentsInfo) SetSegment(segmentID UniqueID, segment *SegmentInfo) {
 
 // SetSegmentIndex sets SegmentIndex with segmentID, perform overwrite if already exists
 func (s *SegmentsInfo) SetSegmentIndex(segmentID UniqueID, segIndex *model.SegmentIndex) {
-	if s.segments[segmentID].segmentIndexes == nil {
-		s.segments[segmentID].segmentIndexes = make(map[UniqueID]*model.SegmentIndex)
+	segment, ok := s.segments[segmentID]
+	if !ok {
+		log.Warn("segment missing for set segment index",
+			zap.Int64("segmentID", segmentID),
+			zap.Int64("indexID", segIndex.IndexID),
+		)
+		return
 	}
-	s.segments[segmentID].segmentIndexes[segIndex.IndexID] = segIndex
+	if segment.segmentIndexes == nil {
+		segment.segmentIndexes = make(map[UniqueID]*model.SegmentIndex)
+	}
+	segment.segmentIndexes[segIndex.IndexID] = segIndex
 }
 
 func (s *SegmentsInfo) DropSegmentIndex(segmentID UniqueID, indexID UniqueID) {

